src/models/decoder_only_pipeline.cpp

```diff
@@ -0,0 +1,229 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "../generators.h"
+#include "decoder_only_pipeline.h"
+
+namespace Generators {
+
+DecoderOnlyPipelineModel::DecoderOnlyPipelineModel(std::unique_ptr<Config> config, OrtEnv& ort_env)
+    : Model{std::move(config)} {
+  bool device_allocator_created = false;
+  for (const auto& model : config_->model.decoder.pipeline) {
+    sessions_.emplace_back(OrtSession::Create(ort_env, (config_->config_path / fs::path(model.filename)).c_str(),
+                                              GetSessionOptions(model.model_id)));
+
+    if (!device_allocator_created && model.session_options.has_value()) {
+      const auto& provider_options = (*model.session_options).provider_options;
+      if (std::any_of(provider_options.begin(), provider_options.end(),
+                      [](const auto& elem) { return !elem.name.empty(); })) {
+        InitDeviceAllocator(*sessions_.back());
```

Comment:
Why not check allocator_device_ for being set vs empty intead of device_allocator_created?